node/src/common/JXCoreEngine.cpp

```diff
@@ -0,0 +1,456 @@
+{
+    try
+    {
+        if (!_started)
+        {
+            LogErrorAndThrow("JXCore engine is not started.");
+        }
+
+        std::function<void(std::string, std::exception_ptr)>* callbackPtr =
+        new std::function<void(std::string, std::exception_ptr)>(callback);
+
+        // The callback function pointer is passed through JavaScript as a hex-formatted number.
+        unsigned long long callId = reinterpret_cast<unsigned long long>(callbackPtr);
+        char callIdBuf[20];
+        snprintf(callIdBuf, sizeof(callIdBuf), "%llx", callId);
+
+        // Create JXValue arguments to the call-script function: callback pointer and script code string.
+        JXValue args[2];
+        JX_New(&args[0]);
+        JX_New(&args[1]);
+        JX_SetString(&args[0], callIdBuf);
+        JX_SetString(&args[1], scriptCode.c_str(), static_cast<int>(scriptCode.size()));
+
+        // Invoke the script function that will evaluate the provided script code then callback
+        // via the result or error callback.
+        JXValue unusedResult;
+        if (JX_CallFunction(reinterpret_cast<JXValue*>(_callScriptFunction), args, 2, &unusedResult))
+        {
+            JX_Free(&unusedResult);
+            LogVerbose("Successfully evaluated script code.");
+        }
+        else
+        {
+            LogErrorAndThrow("Failed to evaluate script code.");
+        }
+
+        JX_Free(&args[0]);
+        JX_Free(&args[1]);
+        JX_Loop();
+    }
+    catch (...)
+    {
+        callback(nullptr, std::current_exception());
+    }
+}
+
+void JXCoreEngine::RegisterCallFromScriptInternal(
+    std::string scriptFunctionName,
+    std::function<void(std::string argsJson)> callback)
+{
+    try
+    {
+        std::function<void(std::string)>* callbackPtr = new std::function<void(std::string)>(callback);
+
+        // The callback function pointer is passed through JavaScript as a hex-formatted number.
+        unsigned long long callId = reinterpret_cast<unsigned long long>(callbackPtr);
+
+        // Note the Array.prototype.slice is necessary for proper array JSON-serialization
+        // because arguments is only an array-like object, not actually an array.
+        const char scriptFunctionFormat[] =
+        "function %s() {"
+            "process.natives.jxcall('%llx', JSON.stringify(Array.prototype.slice.call(arguments)));"
+        "}";
+
+        size_t scriptBufSize = scriptFunctionName.size() + sizeof(scriptFunctionFormat) + 20;
+        std::vector<char> scriptBuf(scriptBufSize);
+        snprintf(scriptBuf.data(), scriptBufSize, scriptFunctionFormat, scriptFunctionName.c_str(), callId);
+
+        // Evaluate the script, which defines the named function as invoking the script call callback.
+        JXValue unusedResult;
+        if (!JX_Evaluate(scriptBuf.data(), nullptr, &unusedResult))
+        {
+            LogErrorAndThrow("Failed to evaluate script callback code.");
+        }
+
+        JX_Loop();
```

Comment:
Why not LoopOnce? if there is anything long running, i.e. user creates an http server, this would never return.